website/static/js/nodesPrivacySettingsTreebeard.js

```diff
@@ -0,0 +1,117 @@
+'use strict';
+
+var $ = require('jquery');
+var m = require('mithril');
+var ko = require('knockout');
+var Treebeard = require('treebeard');
+var $osf = require('js/osfHelpers');
+var projectSettingsTreebeardBase = require('js/projectSettingsTreebeardBase');
+
+
+function expandOnLoad() {
+    var tb = this;  // jshint ignore: line
+    for (var i = 0; i < tb.treeData.children.length; i++) {
+        var parent = tb.treeData.children[i];
+        tb.updateFolder(null, parent);
+        expandChildren(tb, parent.children);
+    }
+}
+
+function expandChildren(tb, children) {
+    var openParent = false;
+    for (var i = 0; i < children.length; i++) {
+        var child = children[i];
+        var parent = children[i].parent();
+        if (child.children.length > 0) {
+            expandChildren(tb, child.children);
+        }
+    }
+    if (openParent) {
+        openAncestors(tb, children[0]);
+    }
+}
+
+function openAncestors (tb, item) {
+    var parent = item.parent();
+    if(parent && parent.id > 0) {
+        tb.updateFolder(null, parent);
+        openAncestors(tb, parent);
+    }
+}
+
+function NodesPrivacyTreebeard(divID, data, nodesState, nodesOriginal) {
```

Comment:
This class has to know a lot about the implementation details of `NodesPrivacy`. For example, `NodesPrivacyTreebeard` needs to know how to update `nodesState`.

You could refactor by encapsulating all `nodesState` updates in `NodesPrivacy`.

```javascript

function NodesPrivacy(...) {

    //...
    new NodesPrivacyTreebeard(..., 
        {
            onClickCheckbox: function(item) {
                // Update item
                item.data.node.is_public = !item.data.node.is_public;
                // update nodesState...
                self.viewModel.nodesState(updatedNodesState);
                // ...
             },
            itemCheck: function(item) { 
                return self.viewModel.nodesState()[item.data.node.id].public;
            }
        })
}

function NodesPrivacyTreebeard(..., options) {
    //...
    self.options = options;
    //...

        resolveRows: function(item){
            //...
            columns.push(
                {
                    data : 'action',
                    sortInclude : false,
                    filter : false,
                    custom : function () {
                        return m('input[type=checkbox]', {
                            disabled : !item.data.node.can_write,
                            onclick : function() {
                                self.options.onClickCheckbox(item);
                            },
                            checked: self.options.itemCheck(item);
                        });
                    }
                },
}
```